### PR TITLE
Add shield iks01a3

### PR DESCRIPTION
**Pull request by avisconti**
x_nucleo_iks01a3 is a companion board that can be plugged through Arduino connectors on top
of Nucleo boards (such as nucleo_f411re).

This PR is adding shield support plus a couple of samples code, one for standard and one for sensorhub mode, plus proper documentation.
The PR is rebased on top of LPS22HH (#13062) and LSM6DSO (#13478) drivers.

### CODE REVIEW

**Comment by erwango** [comment]
@avisconti : I think we can remove tag "platform:STM32". This shield aims to be working on any board supporting arduino header, right?

**Comment by avisconti** [comment]
> @avisconti : I think we can remove tag "platform:STM32". This shield aims to be working on any board supporting arduino header, right?

removed

**Comment by ulfalizer** [comment]
Sorry about that false positive in a comment.

I really want to move ci-tools back into the main Zephyr repo. That way the whitelist could be modified in the same commit, without having to add yet another separate ad-hoc configuration file (which would have to support comments and stuff too).

There will be a pylint whitelist later too.

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: boards/shields/x_nucleo_iks01a3/doc/index.rst [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+Overview
+********
+The X-NUCLEO-IKS01A3 is a motion MEMS and environmental sensor expansion board
+for the STM32 Nucleo. It is equipped with Arduino UNO R3 connector layout, and
+is designed around the LSM6DSO 3D accelerometer and 3D gyroscope, the LIS2MDL
+3-axis magnetometer, the LIS2DW12 3-axis accelerometer, the HTS221 humidity and
+temperature sensor, the LPS22HH pressure sensor, and the STTS751 temperature sensor.
+
+The X-NUCLEO-IKS01A3 interfaces with the STM32 microcontroller via the I²C pin,
```
[the remark this answers is not in the document]
Comment (avisconti) [reply boards/shields/x_nucleo_iks01a3/doc/index.rst]: ACK

File: boards/shields/x_nucleo_iks01a3/doc/index.rst [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+- LIS2DW12 and LSM6DO are on I²C2
+- LIS2MDL, LPS22HH, HTS221 are STTS751 are on I²C1
+
+X-NUCLEO-IKS01A3 board can be configured in two different ways: Mode 1 and Mode 2
+
+
+Mode 1: Standard Mode
+=====================
+
+In standard I²C mode the two bus are connected together. As a consequence, all devices on the shield
```
[the remark this answers is not in the document]
Comment (avisconti) [reply boards/shields/x_nucleo_iks01a3/doc/index.rst]: thx

File: boards/shields/x_nucleo_iks01a3/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+resides on the same I²C bus and are accessible from the main board thru I²C bus.
```
[the remark this answers is not in the document]
Comment (avisconti) [reply boards/shields/x_nucleo_iks01a3/doc/index.rst]: thx

**Comment by avisconti** [comment]
@dbkinder, thx for reviewing the board doc. I have fixed all your requested changes.
Have you taken a look to the two samples docs as well? Are they ok?

**Review by dbkinder — APPROVED** [review]
LGTM, thanks

**Comment by avisconti** [comment]
I re-based the PR on top of the latest changes.
Please note that this PR depends from  #17112 and #17113.

**Comment by avisconti** [comment]
@erwango 
I rebased it on top of master today.
The docuemntation has already been reviewed by dbkinder, but it would be nice if you could
take a look to the actual code itself. It should not take longer as it is quite similar to the other
sensor shields.

File: samples/shields/x_nucleo_iks01a3/sensorhub/src/main.c [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+ * Copyright (c) 2019 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr.h>
+#include <device.h>
+#include <sensor.h>
+#include <stdio.h>
+#include <misc/util.h>
```
Comment (erwango) [inline samples/shields/x_nucleo_iks01a3/sensorhub/src/main.c]: Should be `sys/utils.h` according to on recent: https://github.com/zephyrproject-rtos/zephyr/pull/16540

**Review by erwango — CHANGES_REQUESTED** [review]
Need to review headers inclusion due to recent change.

File: samples/shields/x_nucleo_iks01a3/sensorhub/src/main.c (hunk 4 above)
Comment (avisconti) [reply samples/shields/x_nucleo_iks01a3/sensorhub/src/main.c]: ok, it makes sense.

Comment (avisconti) [reply samples/shields/x_nucleo_iks01a3/sensorhub/src/main.c]: Done. I did it for lps22hh and lsm6dso yesterday. But I forgot to apply the same here.
Thx!

**Review by erwango — APPROVED** [review]

**Comment by avisconti** [comment]
@nashif @galak 
Any further comment from your side?

File: boards/shields/x_nucleo_iks01a3/x_nucleo_iks01a3.overlay [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	hts221@5f {
+		compatible = "st,hts221";
+		reg = <0x5f>;
+		label = "HTS221";
+	};
+
+	lps22hh@5d {
+		compatible = "st,lps22hh";
+		reg = <0x5d>;
+		drdy-gpios = <&gpiob 10 0>;
```
Comment (MaureenHelm) [inline boards/shields/x_nucleo_iks01a3/x_nucleo_iks01a3.overlay]: Can you look at the `arduino_header` introduced in #14057 and commit ce1e990c30f82be8a5c61a41dac1addf317c9538 , and make these gpios not specific to stm32 socs?

File: samples/shields/x_nucleo_iks01a3/sensorhub/README.rst [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+configured for the I2C Arduino connector (both for pin muxing
+and device tree). See for example the :ref:`nucleo_f401re_board` board
+source code:
+
+- :file:`$ZEPHYR_BASE/boards/arm/nucleo_f401re/nucleo_f401re.dts`
+- :file:`$ZEPHYR_BASE/boards/arm/nucleo_f401re/pinmux.c`
+
+Please note that this sample can't be used with boards already supporting
+one of the sensors available on the shield (such as disco_l475_iot1) as zephyr
+does not yet support sensors multiple instances.
```
Comment (MaureenHelm) [inline samples/shields/x_nucleo_iks01a3/sensorhub/README.rst]: This is a little deceiving. Zephyr can certainly support multiple sensor instances, it's the specific drivers that don't actually implement multiple instances.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: samples/shields/x_nucleo_iks01a3/sensorhub/README.rst (hunk 6 above)
Comment (avisconti) [reply samples/shields/x_nucleo_iks01a3/sensorhub/README.rst]: Actually this sentence was copied@paisted from other samples:
   https://docs.zephyrproject.org/latest/samples/shields/x_nucleo_iks01a1/README.html
https://docs.zephyrproject.org/latest/samples/shields/x_nucleo_iks01a2/README.html

But I can change it to "... as sensors multiple instances is not supported".

File: boards/shields/x_nucleo_iks01a3/x_nucleo_iks01a3.overlay (hunk 5 above)
Comment (avisconti) [reply boards/shields/x_nucleo_iks01a3/x_nucleo_iks01a3.overlay]: You mean the drdy gpios? Yes, excellent idea!
I will define arduino_header for nucleo F401re and F411re for testing purpose, and I will
change this shield overlay file to make use of it.

File: samples/shields/x_nucleo_iks01a3/sensorhub/README.rst (hunk 6 above)
Comment (avisconti) [reply samples/shields/x_nucleo_iks01a3/sensorhub/README.rst]: I changed it in this way:

 ```
Please note that this sample can't be used with boards already supporting
 one of the sensors available on the shield (such as disco_l475_iot1)
 as sensors multiple instances are not supported.

```
@MaureenHelm 
Pls check my english and tell me if it is better than before. Thank you.

**Comment by avisconti** [comment]
@MaureenHelm  @erwango 
I added one commit for arduino_header support into nucleo f401re and f411re (it is the same mapping).
Then I changed the shield overlay file in order to have all drdy/irq gpios using the arduino_header.
Then rebased on top of latest master commits and retested both standard and sensorhub samples in both f411 and f401.

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**